File: DFS-BFS-3-changeWord/changeWord4.cpp
#include <bits/stdc++.h>

using namespace std;

/*
dfs
visited

// begin  - - -  hot
//     +++  +++ +++  dot
//    -==-  =-=  =--=-       0=-0=-0  =-0=-0-=   =0-0=-0  dog
//    6787606-0=  cog

//  1 + 1  + 1 +1

//  begin answer =  DFS +1
// hot answer =  DFS +1
// dot answer =  DFS +1
// dog answer =  DFS +1
// cog answer =  DFS +1
*/

#define MAX 50

bool visited[MAX];

int DFS(string begin, string target, vector<string> words)
{
    int count = 0;
    int answer = 0;

    for (int i = 0; i < words.size(); ++i)
    {

        if (!visited[i]) // 방문하지 않았으면
        {
            count = 0;
            for (int j = 0; j < words[i].length(); ++j)
            {
                if (begin[j] == words[i][j])
                    ++count;
            }
            if (count == words[i].length() - 1) // 한 글자만 다른 단어이면
            {
                visited[i] = true;   // 방문처리하고
                begin = words[i];    // begin을 그 단어로 업데이트함
                if (begin == target) // 타겟과 같으면  1 리턴
                    return 1;
                else
                    answer += DFS(begin, target, words) + 1; // 타겟과 다르면  dfs 진입, answer 처리, dfs는 0 or 1 리턴
            }
        }
    }

    return answer;
}

int solution(string begin, string target, vector<string> words)
{
    int answer = 0;

    for (int i = 0; i < words.size(); ++i)
    {
        if (words[i] == target) //  words에 타겟이 있으면
        {
            answer = DFS(begin, target, words); // dfs 시작하고 answer를 리턴함
        }
    }

    return answer;
}

int main(void)
{
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<int> dis_word_len(3, 10);  // 각 단어의 길이는 3 이상 10 이하이며 모든 단어의 길이는 같음
    uniform_int_distribution<int> dis_words_num(3, 50); // words에는 3개 이상 50개 이하의 단어가 있으면 중복되는 단어는 없음
    uniform_int_distribution<int> dis_alpha(97, 122);   // 알바벳 소문자

    int word_len = dis_word_len(gen);   // 단어의 길이
    int words_num = dis_words_num(gen); // words안에 단어의 갯수
    // words_num = 50;

    // begin 생성
    string begin = "";
    for (int i = 0; i < word_len; i++)
        begin += char(dis_alpha(gen));
    cout << "begin: " << begin << endl;

    // target 생성 (단, begin과 같지 않음)
    string target = "";
    while (true)
    {
        for (int i = 0; i < word_len; i++)
            target += char(dis_alpha(gen));

        if (target != begin)
            break;
    }
    cout << "target: " << target << endl;

    // words 생성 (중복 없음)
    vector<string> words;
    while (true)
    {
        string word = "";
        for (int j = 0; j < word_len; j++)
        {
            word += char(dis_alpha(gen));
        }
        if (find(words.begin(), words.end(), word) == words.end()) // 중복되는 단어 아님
            words.push_back(word);
        if (words.size() == words_num)
            break;
    }
    cout << "words_num: " << words_num << endl;
    cout << "words: " << endl;
    for (int i = 0; i < words_num; i++)
    {
        cout << words[i] << ' ';
    }
    cout << endl;

    begin = "hit";
    target = "cog";
    words = {"hot", "dot", "dog", "lot", "log", "cog"};

    auto start_time = chrono::steady_clock::now();
    int result = 0;
    result = solution(begin, target, words);
    auto end_time = chrono::steady_clock::now();
    chrono::duration<double> elapsed_seconds = end_time - start_time;
    cout << "--- " << elapsed_seconds.count() << " seconds ---\n";
    cout << "result: " << result << endl;

    return 0;
}

File: DFS-BFS-3-changeWord/changeWord4.py
import random
import time
from collections import defaultdict

'''
bfs
해당 단어와 words들의 단어를 비교하여 한 글자만 다른 단어들을 리턴
해당 단계의 큐 사이즈 저장해서 사이즈만큼만 반복 돌면 단계별로 나누고 카운트 가능함
from collections import defaultdict
사전에 해당 키값이 없으면  키의 값이 False를 갖도록 추가
해당 단어를 방문처리
다음 단어를 방문하지 않았다면 방문함

깔끔한 버전 chageword6.py
'''


def nextWord(cur, words):  # 해당 단어와 words들의 단어를 비교하여 한 글자만 다른 단어들을 리턴
    ret = []
    for word in words:
        cnt = 0
        for idx in range(len(word)):
            if word[idx] == cur[idx]:
                cnt += 1
        if cnt == len(cur)-1:
            ret.append(word)
    return ret


def bfs(begin, target, words):
    # 사전에 해당 키값이 없으면  키의 값이 False를 갖도록 추가
    visited = defaultdict(lambda: False)
    queue = nextWord(begin, words)
    count = 0
    min = 1e9

    while(len(queue) > 0):  # 큐 빌때 까지 반복
        size = len(queue)  # 해당 단계의 큐 사이즈 저장
        count += 1         # 단계 별로 카운트

        for _ in range(size):   # 위 사이즈 만큼 큐의 단어들을 한 단어씩 반복
            key = queue.pop(0)   # 큐 동작이나 복사 시간 들 듯, deque 이용하자
            visited[key] = True    # 단어를 방문처리
            if (key == target and count < min):    # 타켓과 같으면 카운트 값 업데이트
                min = count
            # 해당 단어와 한 글자씩 다른 다음 단계 단어들 중에서 반복
            for candidate in nextWord(key, words):
                if (visited[candidate] == False):   # 다음 단어를 방문하지 않았다면 방문함
                    queue.append(candidate)

    if min == 1e9:
        return 0
    else:
        return min


def solution(begin, target, words):
    answer = bfs(begin, target, words)
    return answer


len_word = random.randint(3, 10)  # 각 단어의 길이는 3 이상 10 이하이며 모든 단어의 길이는 같음
words_num = random.randint(3, 50)  # words에는 3 ~ 50 개 단어가 있으며 중복 없음
words_num = 50
print("len_word: ", len_word)

# 각 단어는 길이가 모두 같고 알파벳 소문자로 구성됨 (26개 아스키 97 ~ 122)
# begin 만들기
begin = ""
for _ in range(len_word):
    begin += chr(random.randint(97, 122))
print("begin: ", begin)

# target 만들기 (단, begin과 같으면 안됨)
target = ""
while True:
    for _ in range(len_word):
        target += chr(random.randint(97, 122))
    if target != begin:
        break
print("target: ", target)

# words 만들기, 중복되는 단어 없음
words = []
cnt = 0
while True:
    word = ""
    for _ in range(len_word):
        word += chr(random.randint(97, 122))
    # print(len_word, word)
    if word in words:  # 단어가 중복되면 다시 단어 생성함
        continue
    words.append(word)
    cnt += 1
    if cnt == words_num:
        break
print(words_num, words)

# begin = "hit"
# target = "cog"
# words = ["hot", "dot", "dog", "lot", "log", "cog"]


# 실행 시간 및 결과 보기
start_time = time.time()
result = solution(begin, target, words)
print("--- %s seconds ---" % (time.time() - start_time))
print(result)
